File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        
        for i in range(1, len(nums)):
            res[i] = nums[i - 1] * res[i - 1]

        right = nums[-1]

        for i in range(len(nums) - 2, -1, -1):
            res[i] *= right 
            right *= nums[i]
        return res
        

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        res = new ArrayList<>();
        backtrack(nums, 0, target, new ArrayList<>());
        return res;
    }
    void backtrack(int[] nums,int idx, int target, List<Integer> path){
        if (target < 0) return;
        if( target == 0){
            res.add(new ArrayList<>(path));
            return;
        }
        for(int i = idx; i < nums.length; i++){
            path.add(nums[i]);
            backtrack(nums, i, target - nums[i], path);
            path.removeLast();
        }
    }
}
